src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala

```diff
@@ -63,99 +63,45 @@ object BlockMatrix {
     a.map4(b, c, d, f)
 
   def map2(f: (Double, Double) => Double)(l: M, r: M): M =
-    l.map2(r, f)
-
-  private class PairWriter(var i: Int, var j: Int) extends Writable {
-    def this() {
-      this(0, 0)
-    }
-
-    def write(out: DataOutput) {
-      out.writeInt(i)
-      out.writeInt(j)
-    }
-
-    def readFields(in: DataInput) {
-      i = in.readInt()
-      j = in.readInt()
-    }
-  }
-
-  private class MatrixWriter(var rows: Int, var cols: Int, var a: Array[Double]) extends Writable {
-    def this() {
-      this(0, 0, null)
-    }
-
-    def write(out: DataOutput) {
-      out.writeInt(rows)
-      out.writeInt(cols)
-      var i = 0
-      while (i < rows * cols) {
-        out.writeDouble(a(i))
-        i += 1
-      }
-    }
-
-    def readFields(in: DataInput) {
-      rows = in.readInt()
-      cols = in.readInt()
-      a = new Array[Double](rows * cols)
-      var i = 0
-      while (i < rows * cols) {
-        a(i) = in.readDouble()
-        i += 1
-      }
-    }
-
-    def toDenseMatrix(): BDM[Double] = {
-      new BDM[Double](rows, cols, a)
-    }
-  }
-
+    l.map2(r, f)  
+  
   private val metadataRelativePath = "/metadata.json"
-  private val matrixRelativePath = "/matrix"
-
-  /**
-    * Writes the matrix {@code m} to a Hadoop sequence file at location {@code
-    * uri}.
-    *
-    **/
-  def write(dm: M, uri: String) {
-    val hadoop = dm.blocks.sparkContext.hadoopConfiguration
-    hadoop.mkDir(uri)
-
-    dm.blocks.map { case ((i, j), lm) =>
-      (new PairWriter(i, j), new MatrixWriter(lm.rows, lm.cols, lm.toArray))
-    }
-      .saveAsSequenceFile(uri + matrixRelativePath)
-
-    hadoop.writeDataFile(uri + metadataRelativePath) { os =>
-      jackson.Serialization.write(
-        BlockMatrixMetadata(dm.blockSize, dm.rows, dm.cols),
-        os)
-    }
-  }
-
+  
   /**
-    * Reads a BlockMatrix matrix written by {@code write} at location {@code
-    * uri}.
-    *
+    * Reads a BlockMatrix matrix written by {@code write} at location {@code uri}.
     **/
   def read(hc: HailContext, uri: String): M = {
     val hadoop = hc.hadoopConf
     hadoop.mkDir(uri)
-
-    val blocks = hc.sc.sequenceFile[PairWriter, MatrixWriter](uri + matrixRelativePath).map { case (pw, mw) =>
-      ((pw.i, pw.j), mw.toDenseMatrix())
-    }
-
-    val BlockMatrixMetadata(blockSize, rows, cols) =
-      hadoop.readTextFile(uri + metadataRelativePath) { isr =>
+    
+    val BlockMatrixMetadata(blockSize, rows, cols, transposed) =
+      hadoop.readTextFile(uri + metadataRelativePath) { isr  =>
         jackson.Serialization.read[BlockMatrixMetadata](isr)
       }
-
-    new BlockMatrix(blocks.partitionBy(GridPartitioner(rows, cols, blockSize)), blockSize, rows, cols)
-  }
+    
+    val blockRows: Int = ((rows - 1) / blockSize).toInt + 1
```

Comment:
This is duplicating functionality already defined in `GridPartitioner`.  Hoist `gp` up here and use `rowPartitions`, `colPartitions`.